Support PEP-567 context variables

fantix opened the pull request:
Fixes: #5615

### Description

Invoke the given function within a copy of the current PEP-567 context in `greenlet_spawn()`, so that the following sync methods could retrieve the correct context variable.

This PR doesn't cover Python < 3.7 or greenlet < 0.4.17.

However, this PR cannot fix another issue when the context variable is set in the sync method (new greenlet created by `greenlet_spawn()`), the direct async method cannot retrieve the correct context variable, for the same reason why it worked in "Additional Context" of #5615 - the direct async method is actually called from the parent greenlet, whose PEP-567 context doesn't have the variable set:

```python
import asyncio
from contextvars import ContextVar

from sqlalchemy import event
from sqlalchemy.ext.asyncio import create_async_engine
from sqlalchemy.util import await_only

var = ContextVar("var", default="default")


async def async_on_connect():
    print(var.get())


def on_connect(dbapi_connection, connection_record):
    var.set("expected")
    await_only(async_on_connect())


async def main():
    e = create_async_engine("postgresql+asyncpg:///")
    event.listen(e.sync_engine, "connect", on_connect)

    async with e.connect():
        pass


asyncio.run(main())
```

(mind where the context variable is set)

**Expected**: `expected`, **Actual**:  `default`.

A workaround is to make the async method indirect, for example, make it a new task:

```python
def on_connect(dbapi_connection, connection_record):
    var.set("expected")
    await_only(asyncio.ensure_future(async_on_connect()))
```

Expected: `expected`, Actual: `expected`.


### Checklist

This pull request is:

- [x] A short code fix

**Have a nice day!**

fantix committed:
Support PEP-567 context variables

Fixes: #5615

fantix commented on lib/sqlalchemy/util/_concurrency_py3k.py:
```diff
@@ -14,9 +20,26 @@
-        def __init__(self, fn, driver):
-            greenlet.greenlet.__init__(self, fn, driver)
-            self.driver = driver
+        if copy_context is None or greenlet.__version__ < "0.4.17":
```
If we could enforce the required version of greenlet to be 1.0 or above, we could remove the branches here and keep the last branch only.

zzzeek commented on lib/sqlalchemy/util/_concurrency_py3k.py:
```diff
@@ -14,9 +20,26 @@
+                greenlet.greenlet.__init__(self, fn, driver)
+                self.driver = driver
+
+        elif greenlet.__version__ == "0.4.17":
```
can you fill me in on what is specific about 0.4.17 here?   I would think that we would disregard using new behavior for this specific version.   also version comparison would need to be done more robustly, however I would hope that we can use a simple duck-typing approach:

```
    if hasattr(self, "gr_context"):
        self.gr_context = copy_context()

```

fantix replied:
Oh sure - greenlet 0.4.17 included python-greenlet/greenlet#172, which starts new greenlets with empty contexts. #5615 is only caused by greenlet 0.4.17.

I think the duck-typing approach would work fine. I'll update this PR as a draft proposal.

CaselIT replied:
Personally I think it would be fine to constraint the install to `gr_context != 0.4.17`

auvipy reviewed (changes requested):
could you please fix the conflicts?

fantix commented:
Sure thing!

CaselIT commented on test/base/test_concurrency_py3k.py:
```diff
@@ -101,3 +107,36 @@ def go():
+
+    @async_test
+    @testing.requires.python37
+    @testing.skip_if(lambda: not hasattr(greenlet, "gr_context"))
```
shouldn't this work in both cases, since we are blocking the 0.4.17 release?

fantix replied:
Ah, this is because we removed the code to support greenlet < 0.4.17, and this test didn't cover the part when it's not working with <0.4.17, so I added this skip to avoid a false "PASS".

CaselIT replied:
well technically, looking at the test, it just verifies that the context vars are properly set when using `await_only` and `greenlet_spawn` so I think it's not a false PASS for the older greenlet. It just happened to handle it that way without any other code change.

I think we could make it run so we don't accidentally break it for some other reason

fantix replied:
That makes sense, PR is updated with the latest master now 🎉 thanks for reviewing!

fantix committed:
CRF: remove skip to cover <0.4.17

Older greenlet it was just working incidentally,
but we're still testing that nothing else breaks it

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision dcf42f7b78ef3e983fda17f7190cda7b4511e3b4 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
New Gerrit review created for change dcf42f7b78ef3e983fda17f7190cda7b4511e3b4: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2376

sqla-tester commented:
**Federico Caselli** (CaselIT) wrote:

Code-Review+1 Workflow+1

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2376

sqla-tester commented:
**Federico Caselli** (CaselIT) wrote:

mike you can take a look

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2376

sqla-tester commented on test/base/test_concurrency_py3k.py:
```diff
@@ -101,3 +107,35 @@ def go():
+
+        async def async_inner(val):
+            eq_(val, var.get())
+
```
**mike bayer** (zzzeek) wrote:

I tried to make this test simpler and easier to understand.  if I run the latest git master of greenlet, the test passes unless I take out the new contextvar logic in which case it fails, which suggests that this test is accurately testing the behavior.  if you can confirm i am not missing anything that would be helpful.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2376

sqla-tester replied:
**Fantix King** (fantix) wrote:

Yes excactly, that's been my way to verify that the test is working correctly, by commenting out the "self.gr_context = _copy_context()" in the greenlet subclass and watch the test fails.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2376

sqla-tester reviewed (commented):
**mike bayer** (zzzeek) wrote:

Code-Review+2 Workflow+1

(1 comment)

@caselit @fantix one more review as I tried to simplify the test.

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2376

sqla-tester commented:
**Fantix King** (fantix) wrote:

(1 comment)

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2376

sqla-tester commented:
**Federico Caselli** (CaselIT) wrote:

Code-Review+1 Workflow+2

looks good to me

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2376

sqla-tester commented:
**mike bayer** (zzzeek) wrote:

Workflow+2

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2376

sqla-tester commented:
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2376 has been **merged**. Congratulations! :)
